rfc/system/5575-supporting-yunikorn-and-kueue.md

```diff
@@ -0,0 +1,131 @@
+# [Newbie] Supporting Yunikorn and Kueue
+
+**Authors:**
+
+- @yuteng
+
+## 1 Executive Summary
+
+Providing kubernetes (k8s) resource management, gang scheduling and preemption for flyte applications by third-party software, including Apache Yunikorn and Kueue.
+
+## 2 Motivation
+
+Flyte support multi-tenancy and various k8s plugins.
+
+Kueue and Yunikorn support gang scheduling and preemption.
+Gang scheduling guarantees the availability of certain K8s crd services, such as Spark, Ray, with sufficient resource and preemption make sure high priority task execute immediately.
+
+Flyte doesn't provide resource management for multi-tenancy, which hierarchical resource queues of Yunikorn can solve.
+
+## 3 Proposed Implementation
+
+```yaml
+queueconfig:
+  scheduler: yunikorn
+  jobs:
+    - type: "ray"
+      gangscheduling: "placeholderTimeoutInSeconds=60 gangSchedulingStyle=hard"
+    - type: "spark"
+      gangscheduling: "placeholderTimeoutInSeconds=30 gangSchedulingStyle=hard"
+```
+
+Mentioned configuration indicates what queues exist for an org.
+Hierarchical queues will be structured as follows.
+root.org1.ray、root.org1.spark and root.org1.default".
+
+ResourceFlavor allocates resource based on labels which indicates that category-based resource allocation by organization label is available.
```

Comment:
Could you please explain how the the reseource flavor will be determined? Is there a way to derive this from the task decorator args `@task(resources=..., accelerator=...)?

It would be really nice if tasks that need e.g. an A100 GPU were automatically not in the same queue as tasks that need 2 x T4 GPUs. We're using kubeflow pytorch jobs with [scheduler plugins' gang scheduling](https://www.kubeflow.org/docs/components/training/user-guides/job-scheduling/#scheduler-plugins-with-coscheduling) and have observed jobs being starved that the cluster would have had resources for because other jobs couldn't be scheduled but had a higher priority.

Reply:
No, need to create Kueue CRDs first.
A [cluster queue](https://kueue.sigs.k8s.io/docs/concepts/cluster_queue/) defines resource quota which property is defined by [resource flavors](https://kueue.sigs.k8s.io/docs/concepts/resource_flavor/).
I think creating resource flavors to categorizing resources under a cluster queue is available solution.
```
apiVersion: kueue.x-k8s.io/v1beta1
kind: ResourceFlavor
metadata:
  name: "spot-t4"
spec:
  nodeLabels:
    cloud.google.com/gke-accelerator: nvidia-tesla-t4
  nodeTaints:
  - effect: NoSchedule
    key: cloud.google.com/gke-accelerator: nvidia-tesla-t4
    value: "true"
  tolerations:
  - key: "spot-taint"
    operator: "Exists"
    effect: "NoSchedule"
    
   apiVersion: kueue.x-k8s.io/v1beta1
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  namespaceSelector: {} # match all.
  resourceGroups:
  - coveredResources: ["cpu", "memory", "nvidia.com/gpu"]
    flavors:
    - name: "spot-t4"
      resources:
      - name: "cpu"
        nominalQuota: 9
      - name: "memory"
        nominalQuota: 36Gi
      - name: "nvidia.com/gpuu"
        nominalQuota: 50
    - name: "spot-a100"
      resources:
      - name: "cpu"
        nominalQuota: 18
      - name: "memory"
        nominalQuota: 72Gi
      - name: "nvidia.com/gpu"
        nominalQuota: 100
```
In the other hand,  kueue preemption requires Kueue [WorkloadPriorityClass ](https://kueue.sigs.k8s.io/docs/concepts/workload_priority_class/)and patching job with label. 
The plugin received the preemption label and then it should patch it to pods belonging same job